diff --git a/service/java/com/android/server/wifi/WifiAutoJoinController.java b/service/java/com/android/server/wifi/WifiAutoJoinController.java
index e525949..8b5f66b 100644
--- a/service/java/com/android/server/wifi/WifiAutoJoinController.java
+++ b/service/java/com/android/server/wifi/WifiAutoJoinController.java
@@ -61,8 +61,8 @@ public class WifiAutoJoinController {
     private static boolean VDBG = false;
     private static final boolean mStaStaSupported = false;
 
-    public static int mScanResultMaximumAge = 40000; /* milliseconds unit */
-    public static int mScanResultAutoJoinAge = 5000; /* milliseconds unit */
+    public static int mScanResultMaximumAge = 400000000; /* milliseconds unit */
+    public static int mScanResultAutoJoinAge = 500000000; /* milliseconds unit */
 
     private String mCurrentConfigurationKey = null; //used by autojoin
 
@@ -75,7 +75,7 @@ public class WifiAutoJoinController {
     /**
      * Whether to allow connections to untrusted networks.
      */
-    private boolean mAllowUntrustedConnections = false;
+    private boolean mAllowUntrustedConnections = true;
 
     /* For debug purpose only: if the scored override a score */
     boolean didOverride = false;
@@ -1593,24 +1593,7 @@ public class WifiAutoJoinController {
             }
 
             // Try to un-blacklist based on elapsed time
-            if (config.blackListTimestamp > 0) {
-                if (now < config.blackListTimestamp) {
-                    /**
-                     * looks like there was a change in the system clock since we black listed, and
-                     * timestamp is not meaningful anymore, hence lose it.
-                     * this event should be rare enough so that we still want to lose the black list
-                     */
                     config.setAutoJoinStatus(WifiConfiguration.AUTO_JOIN_ENABLED);
-                } else {
-                    if ((now - config.blackListTimestamp) > loseBlackListHardMilli) {
-                        // Reenable it after 18 hours, i.e. next day
-                        config.setAutoJoinStatus(WifiConfiguration.AUTO_JOIN_ENABLED);
-                    } else if ((now - config.blackListTimestamp) > loseBlackListSoftMilli) {
-                        // Lose blacklisting due to bad link
-                        config.setAutoJoinStatus(config.autoJoinStatus - 8);
-                    }
-                }
-            }
 
             if (config.visibility == null) {
                 continue;
diff --git a/service/java/com/android/server/wifi/WifiStateMachine.java b/service/java/com/android/server/wifi/WifiStateMachine.java
index 360bd97..872366a 100644
--- a/service/java/com/android/server/wifi/WifiStateMachine.java
+++ b/service/java/com/android/server/wifi/WifiStateMachine.java
@@ -5812,7 +5812,10 @@ public class WifiStateMachine extends StateMachine implements WifiNative.WifiPno
                     maybeRegisterNetworkFactory();
                     break;
                 case CMD_SCREEN_STATE_CHANGED:
-                    handleScreenStateChanged(message.arg1 != 0);
+                    WifiManager manager = (WifiManager)mContext.getSystemService(Context.WIFI_SERVICE);
+                    if(manager.isWifiEnabled()) {
+                        handleScreenStateChanged(message.arg1 != 0);
+                    }
                     break;
                     /* Discard */
                 case CMD_START_SCAN:
-- 
2.5.0

